### PR TITLE
Use BoTorch's `optimize_acqf` for min-LCB evaluation in optuna terminator

### PR DESCRIPTION
**Pull request by contramundum53**
## Motivation
The original sampling scheme for min-LCB evaluation results in noisy and underestimated regrets, especially when the dimension is high. This causes premature termination and is highly undesirable.

We address this issue by using BoTorch's `optimize_acqf` function.

## Description of the changes
* Change `min_lcb` evaluation algorithm.
* Since BoTorch's `optimize_acqf` function assumes BoTorch's GP model, we entirely remove the abstraction of `BaseGaussianProcess` and simply write all code in `RegretBoundEvaluator`. This simplifies much of the abstraction.

### CODE REVIEW

**Comment by cross32768** [comment]
@contramundum53  Some CIs looks to fail because recently master branch was broken (already fixed). Could you merge master?

File: optuna/terminator/improvement/evaluator.py [hunk 1]
```diff
@@ -59,41 +68,55 @@ def get_preprocessing(self, add_random_inputs: bool = False) -> BasePreprocessin
...
     def evaluate(
         self,
-        trials: List[FrozenTrial],
+        trials: list[FrozenTrial],
         study_direction: StudyDirection,
     ) -> float:
         search_space = IntersectionSearchSpace().calculate(trials)
         self._validate_input(trials, search_space)
 
         fit_trials = self.get_preprocessing().apply(trials, study_direction)
```
Comment (g-votte) [inline optuna/terminator/improvement/evaluator.py:84]:
```suggestion
        trials = self.get_preprocessing().apply(trials, study_direction)
```

How about renaming `fit_trials` to `trials`, since we no longer need to distinguish between `lcb_trials` and `fit_trials`?

There seems to be a bug related to this line actually. Below, in `x, bounds = _convert_trials_to_tensors(trials)`, `trials` are not preprocessed.

**Review by g-votte — CHANGES_REQUESTED** [review]

File: optuna/terminator/improvement/evaluator.py [hunk 2]
```diff
@@ -59,41 +68,55 @@ def get_preprocessing(self, add_random_inputs: bool = False) -> BasePreprocessin
...
 
-        ucb = _min_ucb(trials=fit_trials, gp=self._gp, n_params=n_params, n_trials=n_trials)
-        lcb = _min_lcb(trials=lcb_trials, gp=self._gp, n_params=n_params, n_trials=n_trials)
+        mll = gpytorch.mlls.ExactMarginalLogLikelihood(gp.likelihood, gp)
 
-        regret_bound = ucb - lcb
+        fit_gpytorch_model(mll)
 
-        return regret_bound
+        beta = _get_beta(n_params=len(search_space), n_trials=len(fit_trials))
```
Comment (g-votte) [inline optuna/terminator/improvement/evaluator.py:102]:
```suggestion
        beta = _get_beta(n_params=len(search_space), n_trials=len(trials))
```

If we apply https://github.com/optuna/optuna/pull/4547#discussion_r1152641030, `fit_trials` needs to be renamed here as well.

**Comment by g-votte** [comment]
Let me assign @HideakiImamura as a reviewer.

File: optuna/terminator/improvement/evaluator.py [hunk 3]
```diff
@@ -59,41 +68,55 @@ def get_preprocessing(self, add_random_inputs: bool = False) -> BasePreprocessin
...
+        neg_lcb_func = UpperConfidenceBound(gp, beta=beta, maximize=False)
```
Comment (HideakiImamura) [inline optuna/terminator/improvement/evaluator.py:103]: The negated LCB function and the UCB function are not same, I think. I mean
`- (\mu(x) - \beta \sigma(x)) != \mu(x) + \beta \sigma(x)`

File: tests/terminator_tests/improvement_tests/test_evaluator.py [hunk 4]
```diff
@@ -1,45 +1,99 @@
...
-
-        return mean, std
-
-
 # TODO(g-votte): test the following edge cases
 # TODO(g-votte): - the user specifies non-default top_trials_ratio or min_n_trials
 
 
-def test_evaluate() -> None:
+@pytest.mark.parametrize(
```
Comment (HideakiImamura) [inline tests/terminator_tests/improvement_tests/test_evaluator.py:19]: How about adding test cases to validate the `IntDistribution`, `Float/IntDistribution(log=True)` and `CategoricalDistribution`. This might be worked on the follow-up PRs.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. Basically, LGTM. I have two comments. PTAL.

File: optuna/terminator/improvement/evaluator.py (hunk 3 above)
Comment (contramundum53) [reply optuna/terminator/improvement/evaluator.py]: But we set `maximize=False`, which negates \mu(x). So, in all, this is actually the correct LCB.

Comment (contramundum53) [reply optuna/terminator/improvement/evaluator.py]: I think it's a valid argument that this is not very readable. Do you think adding 20- lines of code to define our own `LowerConfidenceBound` is preferrable?

File: tests/terminator_tests/improvement_tests/test_evaluator.py (hunk 4 above)
Comment (contramundum53) [reply tests/terminator_tests/improvement_tests/test_evaluator.py:23]: Done.

**Comment by contramundum53** [comment]
@HideakiImamura @g-votte PTAL.

File: optuna/terminator/improvement/evaluator.py (hunk 3 above)
Comment (HideakiImamura) [reply optuna/terminator/improvement/evaluator.py]: I see. I misunderstood the minimization and maximization. Looks reasonable.

File: optuna/terminator/improvement/evaluator.py [hunk 5]
```diff
@@ -105,3 +103,101 @@ def _validate_input(
...
+    return gp
+
+
+def _calculate_min_lcb(
+    gp: SingleTaskGP, beta: float, x: torch.Tensor, bounds: torch.Tensor
+) -> float:
+    neg_lcb_func = UpperConfidenceBound(gp, beta=beta, maximize=False)
+
+    with gpytorch.settings.fast_pred_var():  # type: ignore[no-untyped-call]
+        x_opt, lcb = optimize_acqf(
```
Comment (HideakiImamura) [inline optuna/terminator/improvement/evaluator.py:179]: `x_opt` is not used. Can we delete it?

File: tests/terminator_tests/improvement_tests/test_evaluator.py [hunk 6]
```diff
@@ -1,45 +1,52 @@
...
+        ],
+        [
+            create_trial(
+                value=0,
+                distributions={"x": CategoricalDistribution(["a", "b", "c"])},
+                params={"x": "b"},
+            )
+        ],
+    ],
+)
```
Comment (HideakiImamura) [inline tests/terminator_tests/improvement_tests/test_evaluator.py:48]: Could you add the test of `log=True`?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the update. I have two minor comments. PTAL.

File: optuna/terminator/improvement/evaluator.py [hunk 7]
```diff
@@ -105,3 +103,101 @@ def _validate_input(
...
+        )
+        min_lcb = -lcb.item()
+        min_lcb_x = torch.min(-neg_lcb_func(x[:, None, :])).item()
+        min_lcb = min(min_lcb, min_lcb_x)
+
+    return min_lcb
+
+
+def _calculate_min_ucb(
+    gp: SingleTaskGP, beta: float, x: torch.Tensor, bounds: torch.Tensor
```
Comment (g-votte) [inline optuna/terminator/improvement/evaluator.py:196]: `bounds` seems to be a redundant argument. Do you keep it for the consistency with `_calculate_min_lcb`?

File: optuna/terminator/improvement/evaluator.py [hunk 8]
```diff
@@ -41,12 +52,10 @@ def evaluate(
 class RegretBoundEvaluator(BaseImprovementEvaluator):
```
Comment (g-votte) [inline optuna/terminator/improvement/evaluator.py:52]: The `add_random_inputs` argument of `get_preprocessing` is no longer necessary. Could you remove it?

```python
def get_preprocessing(self, add_random_inputs: bool = False) -> BasePreprocessing:
```

Sorry for my commenting here as I cannot directly comment to the line due to the GitHub UI.

File: optuna/terminator/improvement/evaluator.py [hunk 9]
```diff
@@ -41,12 +52,10 @@ def evaluate(
...
     def __init__(
         self,
-        gp: Optional[BaseGaussianProcess] = None,
         top_trials_ratio: float = DEFAULT_TOP_TRIALS_RATIO,
         min_n_trials: int = DEFAULT_MIN_N_TRIALS,
         min_lcb_n_additional_samples: int = 2000,
```
Comment (g-votte) [inline optuna/terminator/improvement/evaluator.py:57]: `min_lcb_n_additional_samples` is no longer necessary. Could you remove it?

**Review by g-votte — CHANGES_REQUESTED** [review]

File: tests/terminator_tests/improvement_tests/test_evaluator.py (hunk 6 above)
Comment (contramundum53) [reply tests/terminator_tests/improvement_tests/test_evaluator.py]: We only need to test IntDistribution and FloatDistribution because all other cases are reduced to these cases in preprocessing.

File: tests/terminator_tests/improvement_tests/test_evaluator.py [hunk 10]
```diff
@@ -1,63 +1,74 @@
...
-        create_trial(
-            value=0,
-            distributions={"a": FloatDistribution(-1.0, 1.0)},
-            params={"a": 0.0},
-        )
-    ]
-
+@pytest.mark.parametrize(
+    # We only need to test FloatDistribution and IntDistribution because the preprocessing logic
+    # reduces all distributions to these two types.
```
Comment (contramundum53) [inline tests/terminator_tests/improvement_tests/test_evaluator.py:24]:
```suggestion
```

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Comment by HideakiImamura** [comment]
@contramundum53 Could you resolve the conflicts?

**Comment by g-votte** [comment]
As mentioned in https://github.com/optuna/optuna/pull/4633#issuecomment-1521301543, we will still face a few issues even after merging this PR and #4633, although these changes will reduce the instability of the regret bounds in many use cases. In order to maintain the simplicity of the implementation, let's close these PRs for now and revisit them after devising solutions for the remaining problems.
